### PR TITLE
feat(overlord): allow adding external managers

### PR DESCRIPTION
**Pull request by paul-rodriguez**
This change enables an external application importing Pebble to extend the overlord functionality by adding additional managers.

Example overlord extension:
```go
package ovldext

type OverlordExtension struct {
	fooManager *foostate.Manager
	barManager *barstate.Manager
}

func New() *OverlordExtension {
    return &OverlordExtension{}
}

func (oe *OverlordExtension) ExtraManagers(o *Overlord) ([]overlord.StateManager, error) {
    oe.fooManager, err := foostate.NewManager(o.State(), o.TaskRunner())
    if err != nil {
        return nil, err
    }
    oe.barManager, err = barstate.NewManager(o.State(), o.TaskRunner())
    if err != nil {
        return nil, err
    }
    return []overlord.StateManager{oe.fooManager, oe.barManager}, nil
}

func (oe *OverlordExtension) FooManager() *foostate.Manager { return oe.fooManager }

func (oe *OverlordExtension) BarManager() *barstate.Manager { return oe.barManager }

func Extension(o *overlord.Overlord) *OverlordExtension {
	return o.Extension().(*OverlordExtension)
}
```

Example startup code:
```go
    dopts := daemon.Options{
        Dir:        pebbleDir,
        SocketPath: socketPath,

        // Supply overlord extension
        OverlordExtension: ovldext.New()            
    }
    
    d, err := daemon.New(&dopts)
    if err != nil {
        return err
    }
```

External HTTP request handler example:
```go  
    :  
    // We can now get access to the overlord
    o := d.Overlord()
    
    // Get access to the custom state manager to perform the HTTP request
    fooManager := ovldext.Extension(o).FooManager()
    :
```

### CODE REVIEW

File: internals/daemon/daemon.go [hunk 1]
```diff
@@ -783,7 +794,14 @@ func New(opts *Options) (*Daemon, error) {
...
 
-	ovld, err := overlord.New(opts.Dir, d, opts.ServiceOutput)
+	ManagerGenerators = append(
+		ManagerGenerators,
+		func(o *overlord.Overlord) (overlord.ManagerTag, overlord.StateManager) {
+			commandMgr := cmdstate.NewManager(o.TaskRunner())
+			return CommandManagerTag, commandMgr
+		})
+
+	ovld, err := overlord.New(opts.Dir, d, opts.ServiceOutput, ManagerGenerators)
```
Comment (flotter) [inline internals/daemon/daemon.go:804]: Instead of using this scheme to now add all built-in overlord managers as well as external ones:

1. Why do we not leave the overlord code as is today, with your ```stateengine``` tweak to make the ```taskRunner``` special.

2. Add a daemon access method to get to the overlord instance? Then we can do this:

```
func (d *Daemon) Overlord() *overlord.Overlord {
    return d.overlord
}
```

Then in ```internals/cli/cmd_run.go``` equivalent in an importer project, we can do:

```
        d, err := daemon.New(&dopts)
        if err != nil {
                return err
        }
        
        // Add custom managers
        c := customManager.New(d.Overlord().State(), d.Overlord().TaskRunner(), ...)
        d.Overlord().StateEngine().AddManager(c)
        
        if err := d.Init(); err != nil {
                return err
        }

```

**Review by flotter — COMMENTED** [review]
Some brainstorming ideas for discussion.

File: internals/daemon/daemon.go (hunk 1 above)
Comment (paul-rodriguez) [reply internals/daemon/daemon.go:804]: Yours is a valid way of doing it.
I wanted to avoid exposing the overlord. I don't have a strong preference either way.

File: internals/overlord/overlord.go [hunk 2]
```diff
@@ -381,6 +402,10 @@ func (o *Overlord) SettleObserveBeforeCleanups(timeout time.Duration, beforeClea
 	return o.settle(timeout, beforeCleanups)
 }
 
+func (o *Overlord) GetManagerByTag(tag ManagerTag) StateManager {
+	return o.customManagers[tag]
+}
+
 // State returns the system state managed by the overlord.
 func (o *Overlord) State() *state.State {
 	return o.stateEng.State()
```
[the remark this answers is not in the document]
Comment (paul-rodriguez) [reply internals/overlord/overlord.go:411]: We still need a way to refer to a manager from another manager (or at least, when building the dependent manager). This is what `GetManagerByTag`  is supposed to accomplish.

But it's true that if you build all your custom managers in the same place then you don't strictly need this.

**Review by anpep — CHANGES_REQUESTED** [review]
I don't think the approach taken here is suitable for a public-facing API.

A simpler approach I suggest: having separate `func New(...) (*Overlord, error)` and `func (*Overlord) Init() error` functions, then have a `func (*Overlord) AddManager(StateManager) error` that we can simply call before `o.Init()`. 

What do you think?

**Comment by paul-rodriguez** [comment]
> I don't think the approach taken here is suitable for a public-facing API.
> 
> A simpler approach I suggest: having separate `func New(...) (*Overlord, error)` and `func (*Overlord) Init() error` functions, then have a `func (*Overlord) AddManager(StateManager) error` that we can simply call before `o.Init()`.
> 
> What do you think?

I don't like letting the user of that code be bitten by forgetting Init or calling AddManager after Init. Right now Overlord is more or less RAII, let's keep it that way if possible.

Here the only purpose of Init would be to ensure the TaskRunner is the last manager. With the changes I made to the StateEngine, Init is not necessary. You can AddManager in cmd_run as Fred suggested.

File: internals/overlord/stateengine.go [hunk 3]
```diff
@@ -56,8 +56,24 @@ type StateEngine struct {
 	state   *state.State
 	stopped bool
 	// managers in use
-	mgrLock  sync.Mutex
-	managers []StateManager
+	mgrLock    sync.Mutex
+	managers   []StateManager
+	taskRunner StateManager
```
[the remark this answers is not in the document]
Comment (flotter) [reply internals/overlord/stateengine.go:61]: We discussed this, and just to clarify my preference is to remove the ```orderedManagers()``` method rendering a new list which simply adds the taskRunner at the end, to simply call respective ```taskRunner``` method after iterating through the other managers. We decided to get some feedback from Gustavo and pick his preference.

**Review by flotter — COMMENTED** [review]
There appears to be a linter failure in one of the tests.

File: internals/overlord/overlord.go [hunk 4]
```diff
@@ -427,20 +424,14 @@ func Fake() *Overlord {
...
 	o.runner = state.NewTaskRunner(s)
 	return o
 }
 
-// AddManager adds a manager to a fake overlord. It cannot be used for
-// a normally initialized overlord those are already fully populated.
 func (o *Overlord) AddManager(mgr StateManager) {
-	if o.inited {
-		panic("internal error: cannot add managers to a fully initialized Overlord")
-	}
```
[the remark this answers is not in the document]
Comment (flotter) [reply internals/overlord/overlord.go:443]: We agreed to use the ```o.StateEngine().AddManager(...)``` directly, instead of modifying the ```o.AddManager(...)``` test method (for use with FakeOverlord), for adding managers. This allows us to leave the ```inited``` flag and its usage unchanged for now.

**Review by anpep — APPROVED** [review]
Excellent work on this @paul-rodriguez!

File: .github/workflows/lint.yml [hunk 5]
```diff
@@ -17,7 +17,7 @@ jobs:
         id: lint
         with:
           version: latest
-          args: '-c .github/.golangci.yml --out-format=colored-line-number'
+          args: '-c .github/.golangci.yml --out-format=colored-line-number --timeout 2m0s'
```
Comment (anpep) [inline .github/workflows/lint.yml:20]: Why is this required though?

**Review by benhoyt — APPROVED** [review]
It's a bit awkward, but I understand why the task runner is special, and this seems reasonable to me -- it's definitely nice and simple!

File: .github/workflows/lint.yml (hunk 5 above)
Comment (flotter) [reply .github/workflows/lint.yml:20]: The default appears too little for the linter. Lots of PRs for various projects increase this.

@paul-rodriguez It looks like this PR from @barrettj12 will set it to 5m using the [run configuration](https://golangci-lint.run/usage/configuration/#run-configuration) key ```timeout``` which is directly set in the config file ```.github/.golangci.yml```.

Maybe we should do the same, which means no conflicting attempts setting timeouts. 5m used by Jordan seems future safe?

**Review by flotter — APPROVED** [review]

File: internals/overlord/overlord.go [hunk 6]
```diff
@@ -64,21 +64,42 @@ type Overlord struct {
...
+// handles to hook an external manager into the overlord's environment.
+type ManagerProvider interface {
+	State() *state.State
+	TaskRunner() *state.TaskRunner
+}
+
+// ManagerGenerator is passed to Overlord to create a manager
+// The return value is a key, value pair where the key has to be a unique
+// identifier for the manager being created.
+type ManagerGenerator func(ManagerProvider) (key any, manager StateManager)
```
Comment (flotter) [inline internals/overlord/overlord.go:88]: I think we will have to include returning an error to comply with how each manager instantiation can fail, and how each of these errors are propagated to the ```overlord.New()``` error return.

**Review by flotter — COMMENTED** [review]
Looks great Paul. I think this is perfect for a first round of reviews, including a proposal discussion with Gustavo. I've taken the liberty to update the example in the PR description for a quick highlight of this scheme for understanding the proposal.

**Review by benhoyt — COMMENTED** [review]
It's a little indirect, but I don't see a significantly cleaner way to achieve what we're trying to achieve here. I've made a few naming suggestions and suggested a couple of ways to simplify the args passed to the new-manager func.

File: internals/daemon/daemon_test.go [hunk 7]
```diff
@@ -99,6 +100,41 @@ func (h *fakeHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+		},
+	}
+
+	d, err := New(&Options{
+		Dir:               s.pebbleDir,
+		SocketPath:        s.socketPath,
+		HTTPAddress:       s.httpAddress,
+		ManagerGenerators: generators,
+	})
+	c.Assert(err, IsNil)
```
[the remark this answers is not in the document]
Comment (paul-rodriguez) [reply internals/daemon/daemon_test.go:126]: I don't think this is the case: we merely create a daemon with New, but we don't Init or Start it.

File: internals/overlord/overlord.go [hunk 8]
```diff
@@ -130,12 +151,28 @@ func New(pebbleDir string, restartHandler restart.Handler, serviceOutput io.Writ
...
 
 	return o, nil
 }
 
+func (o *Overlord) tagManager(tag any, mgr StateManager) {
+	o.externalManagers[tag] = mgr
+	o.addManager(mgr)
+}
+
+func (o *Overlord) GetExternalManager(tag any) StateManager {
```
[the remark this answers is not in the document]
Comment (paul-rodriguez) [reply internals/overlord/overlord.go]: It's meant to be "managers external to pebble". But "external to the overlord" is also a valid interpretation.

File: internals/overlord/overlord.go [hunk 9]
```diff
@@ -64,21 +64,42 @@ type Overlord struct {
...
+	logMgr           *logstate.LogManager
+	externalManagers map[any]StateManager
 }
 
-// New creates a new Overlord with all its state managers.
+// ManagerProvider is the interface that ManagerGenerator depends on
+//
+// Overlord implements ManagerProvider as it provides the necessary
+// handles to hook an external manager into the overlord's environment.
+type ManagerProvider interface {
```
[the remark this answers is not in the document]
Comment (paul-rodriguez) [reply internals/overlord/overlord.go]: I went with 1. but then realized I was creating boilerplate code implementing those functions when I could just be using the ones of Overlord directly.

The reason we're not going with 3. is to avoid exposing the whole Overlord, something which Gustavo insisted on after seeing the previous version.

File: .github/.golangci.yml [hunk 10]
```diff
@@ -11,4 +11,8 @@ linters-settings:
 issues:
   # these values ensure that all issues will be surfaced
   max-issues-per-linter: 0
-  max-same-issues: 0
\ No newline at end of file
+  max-same-issues: 0
+
+run:
+  timeout: 5m
```
[the remark this answers is not in the document]
Comment (paul-rodriguez) [reply .github/.golangci.yml:17]: In my own testing it's more like two minutes, but yeah this is suspicious.

[the remark this answers is not in the document]
Comment (paul-rodriguez) [reply .github/.golangci.yml:17]: Created issue https://github.com/canonical/pebble/issues/286

**Comment by flotter** [comment]
Current status: Further discussions required. Current design unlikely to progress in its current form.

File: internals/daemon/daemon_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (benhoyt) [reply internals/daemon/daemon_test.go:126]: Yeah, you're probably right. However, currently it calls `overlord.New`, which calls `servstate.NewManager`, which starts the (global) reaper with `reaper.Start()`. However, the right fix here is probably to not do that -- which it looks like @flotter is proposing anyway for different reasons in https://github.com/canonical/pebble/issues/284

So probably leave this as is for now.

File: internals/daemon/daemon.go [hunk 11]
```diff
@@ -71,6 +71,10 @@ type Options struct {
 	// ServiceOuput is an optional io.Writer for the service log output, if set, all services
 	// log output will be written to the writer.
 	ServiceOutput io.Writer
+
+	// OverlordExtension is an optional struct used to extend the capabilities
```
Comment (flotter) [inline internals/daemon/daemon.go:75]: Super nitty comment: I would refer to it as an interface, not struct. Also optional does not exactly explain how to not use it. Here are some thoughts for inspiration.
```
// OverlordExtension is an interface allowing the extension of the default Overlord. If no extension is needed, nil should be supplied.
```

File: internals/overlord/overlord.go [hunk 12]
```diff
@@ -70,15 +76,38 @@ type Overlord struct {
 	commandMgr *cmdstate.CommandManager
 	checkMgr   *checkstate.CheckManager
 	logMgr     *logstate.LogManager
+
+	// Overlord extension
```
Comment (flotter) [inline internals/overlord/overlord.go:80]: I think the comment can be dropped. It seems quite clear from the type.

File: internals/overlord/overlord.go [hunk 13]
```diff
@@ -70,15 +76,38 @@ type Overlord struct {
...
+type ManagerEnvironment interface {
+	State() *state.State
+	TaskRunner() *state.TaskRunner
 }
 
-// New creates a new Overlord with all its state managers.
+// NewManagerFunc is passed to Overlord to create a manager
+// The return value is a key, value pair where the key has to be a unique
+// identifier for the manager being created.
+type NewManagerFunc func(ManagerEnvironment) (key any, manager StateManager, err error)
```
Comment (flotter) [inline internals/overlord/overlord.go]: Leftover from previous proposal. Can be deleted I believe ?

File: internals/overlord/overlord.go [hunk 14]
```diff
@@ -70,15 +76,38 @@ type Overlord struct {
...
+
+// New creates a  Overlord with all its state managers.
```
Comment (flotter) [inline internals/overlord/overlord.go:84]: nit: typo: ```an Overlord```

File: internals/overlord/overlord.go [hunk 15]
```diff
@@ -130,12 +159,28 @@ func New(pebbleDir string, restartHandler restart.Handler, serviceOutput io.Writ
...
 	o.checkMgr.NotifyCheckFailed(o.serviceMgr.CheckFailed)
 
-	// the shared task runner should be added last!
+	if extension != nil {
+		extraManagers, err := o.extension.ExtraManagers(o)
+		if err != nil {
+			return nil, err
+		}
+		for _, manager := range extraManagers {
+			o.addManager(manager)
```
Comment (flotter) [inline internals/overlord/overlord.go:154]: nit: I would replace this with ```o.stateEng.AddManager(manager)```. The original purpose of ```addManager``` was to assign manager instances to the overlord struct (snapd), which has been removed. This wrapper adds no value. I think we can clean up (delete the method) the code as part of this?

**Review by flotter — COMMENTED** [review]
Great! Added a few nitty and minor comments, as I think we may have time to do a detailed code review during our next meeting.

File: internals/overlord/overlord.go (hunk 15 above)
Comment (paul-rodriguez) [reply internals/overlord/overlord.go:154]: I wouldn't delete the `Overlord.addManager` method in this PR, but I will use `stateEng.AddManager` if you think it's better.

File: internals/daemon/daemon.go (hunk 11 above)
Comment (paul-rodriguez) [reply internals/daemon/daemon.go:75]: See the comments of the other fields in the Options struct: optional means "either something or nil" (pretty sure this is a widely used Golang convention). And the default behavior in case of "nil" is not to extend (I think this is obvious).

Comment (flotter) [reply internals/daemon/daemon.go]: OK thanks. Makes sense!

File: internals/daemon/daemon_test.go [hunk 16]
```diff
@@ -99,6 +100,45 @@ func (h *fakeHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+		HTTPAddress:       s.httpAddress,
+		OverlordExtension: &fakeExtension{},
+	})
+	c.Assert(err, IsNil)
+
+	err = d.overlord.StateEngine().Ensure()
+	c.Assert(err, IsNil)
+	extension, ok := d.overlord.Extension().(*fakeExtension)
+	c.Assert(ok, Equals, true)
+	c.Assert(extension, NotNil)
```
Comment (paul-rodriguez) [inline internals/daemon/daemon_test.go:136]: remove this, add bad case test

File: internals/overlord/overlord.go [hunk 17]
```diff
@@ -70,15 +76,23 @@ type Overlord struct {
...
 
-// New creates a new Overlord with all its state managers.
+// New creates an Overlord with all its state managers.
 // It can be provided with an optional restart.Handler.
-func New(pebbleDir string, restartHandler restart.Handler, serviceOutput io.Writer) (*Overlord, error) {
+func New(
+	pebbleDir string,
+	restartHandler restart.Handler,
+	serviceOutput io.Writer,
+	extension Extension) (*Overlord, error) {
```
Comment (paul-rodriguez) [inline internals/overlord/overlord.go:89]: put all args into an options struct

**Review by flotter — APPROVED** [review]
Great work Paul. Looks super nice.

Regarding the PR description example, feel free to remote it or change it if you want. I find it essential assisting reviewers get the complete picture for things talking about some hypothetical external use-case (especially when discussing in person), but do not feel like we have to keep it there, or that version of it.

File: internals/overlord/overlord.go [hunk 18]
```diff
@@ -49,6 +49,25 @@ var (
 	defaultCachedDownloads = 5
 )
 
+// Extension represents an extension of the Overlord.
+type Extension interface {
+	// ExtraManagers is called when building an Overlord.
```
Comment (benhoyt) [inline internals/overlord/overlord.go:54]: I don't think this is quite enough of a comment. Perhaps:

```suggestion
	// ExtraManagers is called when building an Overlord, allowing additional
	// managers to be created and added to the overlord.
```

File: internals/overlord/overlord.go [hunk 19]
```diff
@@ -49,6 +49,25 @@ var (
...
+// Options is the arguments passed to construct an Overlord.
+type Options struct {
+	// PebbleDir is the path to the pebble directory. It must be provided.
+	PebbleDir string
+	// RestartHandler is an optional structure to handle restart requests.
+	RestartHandler restart.Handler
+	// ServiceOutput is an optional output for the logging manager.
+	ServiceOutput io.Writer
+	// Extension is an optional struct defining additional components to add to
+	// the overlord.
```
Comment (benhoyt) [inline internals/overlord/overlord.go:67]: Yeah, the wording here is tricky, but I agree with @flotter it's weird/confusing to say "struct" when this is clearly an interface. (Which could be fulfilled by any type, even a `func` type, or another interface.)

I think the updated comment on `Extension.ExtraManagers` will help, and maybe after that we can change this to:

```suggestion
	// Extension allows extending the overlord with additional managers.
```

**Review by benhoyt — APPROVED** [review]
This seems reasonable to me. Couple of comments about comments.

File: internals/overlord/overlord.go [hunk 20]
```diff
@@ -130,12 +158,28 @@ func New(pebbleDir string, restartHandler restart.Handler, serviceOutput io.Writ
...
 	return o, nil
 }
 
+func (o *Overlord) Extension() Extension {
+	return o.extension
+}
+
 func (o *Overlord) addManager(mgr StateManager) {
 	o.stateEng.AddManager(mgr)
 }
```
[the remark this answers is not in the document]
Comment (paul-rodriguez) [reply internals/overlord/overlord.go:185]: forgot about that, thanks!

File: internals/overlord/overlord.go [hunk 21]
```diff
@@ -130,14 +158,26 @@ func New(pebbleDir string, restartHandler restart.Handler, serviceOutput io.Writ
...
+
+	// TaskRunner must be the last manager added to the StateEngine,
+	// because TaskRunner runs all the tasks required by the managers that ran
+	// before it.
 	o.stateEng.AddManager(o.runner)
 
 	return o, nil
 }
 
-func (o *Overlord) addManager(mgr StateManager) {
```
Comment (anpep) [inline internals/overlord/overlord.go:139]: Yes!

**Review by anpep — APPROVED** [review]
Stellar work on this Paul, thank you!

File: internals/overlord/overlord.go [hunk 22]
```diff
@@ -49,6 +49,25 @@ var (
 	defaultCachedDownloads = 5
 )
 
+// Extension represents an extension of the Overlord.
+type Extension interface {
+	// ExtraManagers is called when building an Overlord, it allows additional
+	// managers to be added to the Overlord.
```
Comment (niemeyer) [inline internals/overlord/overlord.go:55]:
```
// ExtraManagers allows additional StateManagers to be used.
```

Whether this is called when building the overlord or elsewhere is up to the implementation of the overlord.

File: internals/overlord/overlord.go [hunk 23]
```diff
@@ -49,6 +49,25 @@ var (
...
+
+// Options is the arguments passed to construct an Overlord.
+type Options struct {
+	// PebbleDir is the path to the pebble directory. It must be provided.
+	PebbleDir string
+	// RestartHandler is an optional structure to handle restart requests.
+	RestartHandler restart.Handler
+	// ServiceOutput is an optional output for the logging manager.
+	ServiceOutput io.Writer
+	// Extension, if set, defines additional components to add to the overlord.
```
Comment (niemeyer) [inline internals/overlord/overlord.go:67]:
```
// Extension allows extending the overlord with externally defined features.
```

This is what the extension is. The documentation at the moment is describing ExtraManagers, but that's up to the Extension itself to document.

**Review by niemeyer — APPROVED** [review]
Thank you, that turned out well. Only documentation tuning suggested below.

**Comment by flotter** [comment]
Last two corrections added. Approved by Gustavo. Merging.
